File: music-commons-core/src/main/java/net/transgressoft/commons/music/audio/ImmutableAlbum.java
package net.transgressoft.commons.music.audio;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;
import com.neovisionaries.i18n.CountryCode;

import java.util.Optional;

/**
 * @author Octavio Calleya
 */
public class ImmutableAlbum implements Album {

    private final String name;
    private final Artist albumArtist;
    private final boolean isCompilation;
    private final short year;
    private final Label label;
    private final byte[] coverBytes;

    public static final Album UNKNOWN_ALBUM = new ImmutableAlbum("");

    public ImmutableAlbum(String name, Artist albumArtist, boolean isCompilation, short year, Label label, byte[] coverBytes) {
        this.name = beautifyName(name);
        this.albumArtist = albumArtist;
        this.isCompilation = isCompilation;
        this.year = year;
        this.label = label;
        this.coverBytes = coverBytes;
    }

    public ImmutableAlbum(String name, Artist albumArtist, boolean isCompilation, short year, Label label) {
        this.name = beautifyName(name);
        this.albumArtist = albumArtist;
        this.isCompilation = isCompilation;
        this.year = year;
        this.label = label;
        this.coverBytes = null;
    }

    public ImmutableAlbum(String name, short year) {
        this.name = beautifyName(name);
        this.albumArtist = ImmutableArtist.UNKNOWN_ARTIST;
        this.isCompilation = false;
        this.year = year;
        this.label = new ImmutableLabel("", CountryCode.UNDEFINED);
        this.coverBytes = null;
    }

    public ImmutableAlbum(String name) {
        this.name = beautifyName(name);
        this.albumArtist = ImmutableArtist.UNKNOWN_ARTIST;
        this.isCompilation = false;
        this.year = -1;
        this.label = new ImmutableLabel("", CountryCode.UNDEFINED);
        this.coverBytes = null;
    }

    private String beautifyName(String name) {
        return name.replaceAll("\\s+", " ");
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public Artist albumArtist() {
        return albumArtist;
    }

    @Override
    public boolean isCompilation() {
        return isCompilation;
    }

    @Override
    public short year() {
        return year;
    }

    @Override
    public Label label() {
        return label;
    }

    @Override
    public Optional<byte[]> coverImage() {
        return java.util.Objects.equals(coverBytes, new byte[0]) ? Optional.empty() : Optional.of(coverBytes);
    }

    @Override
    public int compareTo(Album o) {
        int nameComparison = name.compareTo(o.name());
        int artistComparison = albumArtist.name().compareTo(o.albumArtist().name());
        int labelComparison = label.name().compareTo(o.label().name());
        int yearComparison = year - o.year();

        if (nameComparison != 0) {
            return nameComparison;
        } else if (artistComparison != 0) {
            return artistComparison;
        } else if (labelComparison != 0) {
            return labelComparison;
        } else return yearComparison;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ImmutableAlbum that = (ImmutableAlbum) o;
        return isCompilation == that.isCompilation &&
                year == that.year &&
                Objects.equal(name, that.name) &&
                Objects.equal(albumArtist, that.albumArtist) &&
                Objects.equal(label, that.label);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name, albumArtist, isCompilation, year, label);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("name", name)
                .add("albumArtist", albumArtist)
                .add("isCompilation", isCompilation)
                .add("year", year)
                .add("label", label)
                .toString();
    }
}

File: music-commons-core/src/main/kotlin/net/transgressoft/commons/music/audio/ImmutableAlbum.kt
/******************************************************************************
 * Copyright (C) 2025  Octavio Calleya Garcia                                 *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify       *
 * it under the terms of the GNU General Public License as published by       *
 * the Free Software Foundation, either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * GNU General Public License for more details.                               *
 *                                                                            *
 * You should have received a copy of the GNU General Public License          *
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.     *
 ******************************************************************************/

package net.transgressoft.commons.music.audio

import net.transgressoft.lirp.entity.ReactiveEntityBase
import mu.KotlinLogging
import java.lang.ref.SoftReference

/**
 * Immutable album holding an ordered list of audio items for a single [AlbumDetails] key.
 *
 * Items are delivered pre-sorted by the projection's [entryOrdering] comparator (disc then track
 * number). The list is stored verbatim — no re-sorting or deduplication happens here because the
 * lirp projection guarantees both ordering and distinctness via its internal reverse index.
 *
 * The [album] field is the derived representative for this bucket (most-frequent field values
 * across all tracks), which preserves original name casing and carries the majority year/label.
 * [uniqueId] is derived from the canonical key (not the representative) so that it remains stable
 * when the representative's year or casing changes — e.g., correcting a single mistag does not
 * change the album's identity in event routing.
 *
 * Cover bytes are resolved lazily from the first track that carries cover data, cached behind a
 * [SoftReference] so that memory pressure can evict the bytes and re-resolution happens
 * transparently on the next access.
 *
 * Documented behavior: `Artist`, `Album`, and `Label` flyweight instances are cached in
 * [SoftReference]-backed maps and may be GC-evicted under memory pressure. After eviction a
 * subsequent lookup creates a new instance that is `equals`-equivalent but not `===`-identical
 * to the prior one. Always use value-based `equals` (or standard collection operations) to compare
 * flyweight instances; reference equality (`===`) is not a stable identity across GC cycles.
 *
 * @param I The type of audio items contained in this album
 */
internal class ImmutableAlbum<I>(override val album: AlbumDetails, tracks: List<I>) :
    Album<I>,
    ReactiveAlbum<Album<I>, I>,
    Comparable<Album<I>>,
    ReactiveEntityBase<AlbumDetails, Album<I>>()
    where I : ReactiveAudioItem<I> {

    private val logger = KotlinLogging.logger {}

    // Defensive copy; lirp delivers the list pre-sorted so no re-sorting is needed.
    override val tracks: List<I> = tracks.toList()

    override val id: AlbumDetails = album

    override val uniqueId: String = album.canonicalKey().id()

    override val isEmpty: Boolean
        get() = tracks.isEmpty()

    override val size: Int
        get() = tracks.size

    @Volatile
    private var coverRef: SoftReference<ByteArray>? = null

    // `noCover` means "no cover was found at the last load attempt".
    // It is reset whenever `coverRef` had been set (a cover was found previously) and is then
    // evicted by the GC — indicated by a non-null `coverRef` whose referent is null — so that
    // the next access retries loading from the tracks rather than returning null permanently.
    @Volatile
    private var noCover: Boolean = false

    override val coverImageBytes: ByteArray?
        get() {
            val currentRef = coverRef
            val cached = currentRef?.get()
            if (cached != null) return cached
            // If coverRef was previously set (a cover was loaded) but has since been evicted,
            // reset noCover so the next synchronized block retries loading.
            if (currentRef != null && cached == null) noCover = false
            if (noCover) return null
            synchronized(this) {
                val lockedRef = coverRef
                lockedRef?.get()?.let { return it }
                // Inner eviction check: re-clear noCover if coverRef was evicted while waiting.
                if (lockedRef != null && lockedRef.get() == null) noCover = false
                if (noCover) return null
                val bytes = firstCoverImageBytes(tracks)
                return if (bytes != null) {
                    coverRef = SoftReference(bytes)
                    bytes
                } else {
                    noCover = true
                    null
                }
            }
        }

    override fun compareTo(other: Album<I>): Int = this.album.compareTo(other.album)

    override fun clone(): ImmutableAlbum<I> = this

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as ImmutableAlbum<*>
        if (album != other.album) return false
        return tracks == other.tracks
    }

    override fun hashCode(): Int = 31 * album.hashCode() + tracks.hashCode()

    override fun toString() = "ImmutableAlbum(album=$album, size=$size)"

    init {
        logger.trace { "Album created for ${album.name}" }
    }
}